Lua base64 methods

saurtron opened the pull request:
### Work done

- Add `Encoding.EncodeBase64`, `Encoding.DecodeBase64` and  `Encoding.IsValidBase64`
- Add `Encoding.EncodeBase64url`, `Encoding.DecodeBase64url` and `Encoding.IsValidBase64url`

#### Related issues

- https://github.com/beyond-all-reason/RecoilEngine/issues/877

### Testing

- use https://github.com/saurtron/Beyond-All-Reason/tree/test-base64

### Remarks

- Also introduce a mechanism to include modules common to all lua environments.
  - Didn't see anything for this, and later the common ones can be refactored into that, like `LuaConstEngine`.
  - Maybe base64/encoding isn't wanted everywhere, but I don't see why not.
- Could add to `string` instead of `Encoding`, or some other namespace, but then I think the names would need to conform to the string module casing.

sprunk commented:
> Thought about using a different table than Spring, actually would rather do that, but did it like this to conform with the above issue for now.

Perhaps put this in `string.` since it's just text both ways? There's precedent for modifying standard Lua libraries with LuaMathExtra.cpp and LuaTableExtra.cpp.

sprunk commented:
Another remark. There's a handful of extra variants of base64 (e.g. `base64url`). In Recoil games, base64 is mostly used to cram data into modoptions, and the startscript format (which is how modoptions are passed) doesn't work with some special chars like `;` and `=`. It would be good if whichever variant this function uses worked with modoptions out of the box, or at least if it was available as an alternative.

badosu commented:
I really would like for us to start offloading scope from `Spring` or other similarly overloaded namespaces, at least for new callouts.

For example, not my specific proposal but an example of the benefits to this approach:

- `Base64.{encode,decode,encodeUrl,DecodeUrl`: much easier to reason about and inspect documentation for
- Easy to just know "Base64 exists in this lua env, does not in this" than "Spring.{X1,X2,X3,X4} exists in this lua env, not in this, just like Spring.X(5...N)"

As for monkey patching `string`, I'm not qualified to consider the wider implications of this change for runtime characteristics (does it impact performance on table lookup, etc) but I can see the appeal for developer ergonomics (`"asd":EncodeBase64()`, `decoded = someString:DecodeBase64()`). I won't chime in on this particular choice.

saurtron commented:
> I really would like for us to start offloading scope from `Spring` or other similarly overloaded namespaces, at least for new callouts.
> 
> For example, not my specific proposal but an example of the benefits to this approach:
> 
>     * `Base64.{Encode,Decode,EncodeUrl,DecodeUrl}`: much easier to reason about and inspect documentation for
> 
>     * Easy to just know "Base64 exists in this lua env, does not in this" than "Spring.{X1,X2,X3,X4} exists in this lua env, not in this, just like Spring.X(5...N)"

I'm 100% on this one about not putting yet more methods into `Spring`.

I wouldn't do it under top level Base64 tho (too few functions imo, and with this approach we'd end with too many top level pretty empty modules). Not completely against the idea either.

Maybe a top level `Encoding` module?. Otherwise I guess inside `string` should be fine, it may have a small perf implications like you say, but those can be avoided by having a speedup `local Base64Encode = string.Base64Encode`, for cases where perf don't matter it does allow `myString:Base64Encode()` notation which is nicer although kind of niche in lua atm due to perf implications.


> Is there a reason we'd consider undesirable for engine to provide a "blessed" lua-based base64 implementation

Yeah, for performance it's much better to have it in c++, also it's a standard so not much games need to customize there

badosu commented:
`Encoding` or any name that encapsulates the need for this concern (json support would go there too I think) sounds good to me, as for `string` again I'd have to defer to others.

saurtron commented:
> `Encoding` or any name that encapsulates the need for this concern (json support would go there too I think) sounds good to me, as for `string` again I'd have to defer to others. Btw, I just ask you to confer whether you are exposing this API to `gamedata/defs.lua` or any other relevant lua entrypoint consumers (there are others than the main ones).

Actually, thanks to your remark noticed I was missing including it inside LuaParser so it will be accessible inside defs.lua. Unsure about whether there are other places to include at, but looking for IsEngineMinVersion that should also be everywhere I don't think so.

saurtron commented:
I have added Base64url methods. Not sure they are needed atm, but I guess someone might want to use them at some point.

Also added an optional stripPadding bool argument to base64 encoding, as I think that's what modrules requires. Made stripPadding default to false, since strictly that's "real base64", although for practical purposes we can change it to default to true, maybe.

Also changed the module where methods live, to `Encoding`, thought about adding them to `string` but I feel in the end that will make the string module also a bit of a kitchen sink and make them more difficult to locate, also, I feel like if they are added to string, the CamelCasing should be changed, since string module doesn't generally use that casing, then it'd be a bit out of place... all in all I think it's better to have a separate module, then whoever wants can still link them into `string`, to get the syntax sugar.

Only problem with introducing a new module, is games will need to forward it to gadgets/widgets for it to be available, since those get a restricted environment. Actually had to link it in basecontent. We can still change it's location if you prefer, so pls let me know what you think @badosu and @sprunk.

sprunk reviewed (approved):
Would be nice to check if the encoded values survive being pushed through startscript. So something like (in synced):
```lua
local strings = {}
math.randomseed(123) -- for reproducibility
for L = 1, 20 do
for i = 1, 10 do
  local str = ""
  for j = 1, L do
    str = str .. string.char(math.random(0, 255))
  end
  strings:insert(str)
end
-- a bunch of manual tests too
strings:insert("return { foo = bar }")
strings:insert("x = 1.e-6 + 0.123")
strings:insert("works??!??!///")
strings:insert("x=1")
strings:insert("x=1;")
strings:insert("x=1;\n")
strings:insert("x\n=1;\n")
strings:insert(";")
strings:insert("\n")
strings:insert("")

if Spring.GetGameRulesParams("verify_encoding") then
  for i = 1, #strings do
    if Encoding.Base64DecodeUrl(Spring.GetGameRulesParams("encoded" .. i)) ~= strings[i] then error("WRONG") end
  end
else
  local startscript = "[GAME]\n ... blabla player and map ... [MODOPTIONS]\nverify_encoding=1\n"
  for i = 1, #strings do
    startscript = startscript .. "encoded" .. i .. "=" .. Encoding.Base64Encode(strings[i]) .. ";"
  end
  Spring.Reload(startscript)
end
```

I'm giving an approval anyway (regardless of the test above) because if neither of the two versions (default or url) works this way we can just add a third that does.

saurtron commented on rts/Lua/LuaEncoding.cpp:
```diff
@@ -0,0 +1,171 @@
+			case '_': c = '/'; break;
+		}
+	}
+	const std::string decoded = base64_decode(text);
+
+	lua_pushsstring(L, decoded);
+	return 1;
```
The idea is let people validate in advance if they need, with the included methods. Although as said that still doesn't guarantee valid utf8, just valid base64, so not sure what to do here since didn't see a good isvalidutf8 method to reuse for now, and didn't want to go all the way into that for now. Also ppl might be storing binary there instead of utf8.

On the other hand if we do validate and its found to be invalid base64 maybe we should luaL_error instead, and let ppl use the validation method in advance if they want to avoid that. Only issue is then we'd be validating twice unless adding an opt there about whether to validate, and didn't want to do that either to keep it simple.

saurtron replied:
Also worth to mention we could validate while parsing it by modifying the base64 lib we use, and also add better base64url support, but did it like this to avoid modifying vendor files.

saurtron replied:
> Sure, but what if they don't? Consider:
> ```lua
> local probably_garbage = Encoding.Base64Decode("φφφ")
> Spring.Echo(probably_garbage)
> ```

Tested this and the behaviour is: it stops the decode as soon as it finds a non base64 char, and returns whatever it has.

If we want to allow "contaminated" base64 could also add a cleanup method, specially to remove spaces or newlines... although I think those should not be allowed in base64, but they are in some variants.

saurtron commented:
> Would be nice to check if the encoded values survive being pushed through startscript. So something like (in synced):

Will do the test before merging to be sure, but I think it should be fine since I think what breaks the startscript are the "=" signs but otherwise they're simply base64 (always heard ppl talking about removing the "=" symbols and judging from startscript format).

saurtron commented:
> Would be nice to check if the encoded values survive being pushed through startscript. So something like (in synced):

Tested your suggestion, like this: [dbg_test_base64.lua](https://github.com/saurtron/Beyond-All-Reason/blob/2c60da78f0848d08dae98587d32dfacebe76c380/luarules/gadgets/dbg_test_base64.lua).

It does show the format is compatible as is, and most of the manual tests break the format since they have = or other of the reserved chars, probably being `{}[]=;` but not sure.

It did catch something completely unexpected, due to how the input to the encoding functions was being called it was having problems with embedded NULLS, fixed those.

sprunk commented on rts/Lua/LuaEncoding.cpp:
```diff
@@ -0,0 +1,174 @@
+
+
+// For validation
+static const std::string base64Chars =
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "abcdefghijklmnopqrstuvwxyz"
+    "0123456789+/=";
+
+static const std::string base64UrlChars =
+    "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+    "abcdefghijklmnopqrstuvwxyz"
+    "0123456789-_";
```
ws
```suggestion
static const std::string base64Chars =
	"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	"abcdefghijklmnopqrstuvwxyz"
	"0123456789+/=";

static const std::string base64UrlChars =
	"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	"abcdefghijklmnopqrstuvwxyz"
	"0123456789-_";
```

sprunk commented on rts/Lua/LuaEncoding.cpp:
```diff
@@ -0,0 +1,174 @@
+
+/*** Decodes a base64 string
+ *
+ * @function Encoding.DecodeBase64
+ *
+ * @param text string Text to decode
```
```suggestion
 * @function Encoding.DecodeBase64
 *
 * Decodes a base64 encoded string. If it encounters
 * invalid data, it stops and returns whatever it parsed so far.
 *
 * @param text string Text to decode
```
